applications/OptimizationApplication/custom_utilities/control/sigmoidal_projection_utils.cpp

```diff
@@ -0,0 +1,290 @@
+}
+
+double ProjectValueForward(
+        double xValue,
+        std::vector<double> xLimits,
+        std::vector<double> yLimits,
+        double beta,
+        int penalFac)
+{
+
+    size_t size = xLimits.size();
+    double x1 = xLimits[0], x2 = xLimits[1], y1 = yLimits[0], y2 = yLimits[1];
+
+    for (size_t i = 0; i < size - 1; ++i) {
+        if (xValue >= xLimits[i] && xValue <= xLimits[i + 1]) {
+            x1 = xLimits[i];
+            x2 = xLimits[i + 1];
+            y1 = yLimits[i];
+            y2 = yLimits[i + 1];
+            break;
+        }
+    }
+
+    if (xValue > xLimits[size - 1]) {
+        x1 = xLimits[size - 2];
+        x2 = xLimits[size - 1];
+        y1 = yLimits[size - 2];
+        y2 = yLimits[size - 1];
+    }
+
+    double pow_val = -2.0 * beta * (xValue - (x1 + x2) / 2)  * penalFac;
+    double limit = std::log1p(std::numeric_limits<double>::max());
+    pow_val = std::clamp(pow_val, -limit, limit);
+    return (y2 - y1) / std::pow((1.0 + std::exp(pow_val)),penalFac) + y1;
+
+}
+
+double ProjectValueBackward(
+        double yValue,
+        std::vector<double> xLimits,
+        std::vector<double> yLimits,
+        double beta,
+        int penalFac)
+{
+
+    size_t size = xLimits.size();
+    if (yValue > yLimits[size - 1] || yValue < yLimits[0])
+        KRATOS_ERROR << "SigmoidalProjectionUtils::ProjectValueBackward: yValue "
+                     <<yValue<<" is out of the given range "<< yLimits <<"\n";
+
+    if (std::abs(yValue)>std::numeric_limits<double>::max())
+        KRATOS_ERROR << "SigmoidalProjectionUtils::ProjectValueBackward: yValue "
+                     <<yValue<<" is too big for backward projection !!! "<< yLimits <<"\n";
+
+    double x1 = xLimits[0], x2 = xLimits[1], y1 = yLimits[0], y2 = yLimits[1];
+    for (size_t i = 0; i < size - 1; ++i) {
+        if (yValue >= yLimits[i] && yValue <= yLimits[i + 1]) {
+            x1 = xLimits[i];
+            x2 = xLimits[i + 1];
+            y1 = yLimits[i];
+            y2 = yLimits[i + 1];
+            break;
+        }
+    }
+
+    if (std::abs(yValue-y1)<std::numeric_limits<double>::epsilon())
+        return x1;
+    else if (std::abs(yValue-y2)<std::numeric_limits<double>::epsilon())
+        return x2;
+    else
+        return ((x2+x1)/2.0) + (1.0/(-2.0*beta)) * std::log(((y2-y1)/(yValue-y1))-1);
+}
+
+double ComputeFirstDerivativeAtValue(
+        double xValue,
+        std::vector<double> xLimits,
+        std::vector<double> yLimits,
+        double beta,
+        int penalFac)
+{
+
+    size_t size = xLimits.size();
+    double x1 = xLimits[0], x2 = xLimits[1], y1 = yLimits[0], y2 = yLimits[1];
+
+    for (size_t i = 0; i < size - 1; ++i) {
+        if (xValue >= xLimits[i] && xValue <= xLimits[i + 1]) {
+            x1 = xLimits[i];
+            x2 = xLimits[i + 1];
+            y1 = yLimits[i];
+            y2 = yLimits[i + 1];
+            break;
+        }
+    }
+
```

Comment:
Const correctness, passing by reference and `Kratos` variable naming style.

The logic can be simplified as in the above examples to be faster.